utils/src/server/middleware.rs

```diff
@@ -0,0 +1,85 @@
+// Copyright 2019-2021 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any
+// person obtaining a copy of this software and associated
+// documentation files (the "Software"), to deal in the
+// Software without restriction, including without
+// limitation the rights to use, copy, modify, merge,
+// publish, distribute, sublicense, and/or sell copies of
+// the Software, and to permit persons to whom the Software
+// is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice
+// shall be included in all copies or substantial portions
+// of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+// ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
+// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
+// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! TODO
+
+/// TODO
+pub trait Middleware: Send + Sync + Clone + 'static {
+	/// Intended to carry timestamp of a request, for example `std::time::Instant`. How the middleware
+	/// measures time, if at all, is entirely up to the implementation.
+	type Instant: Send + Copy;
+
+	/// Called when a new JSON-RPC comes to the server.
+	fn on_request(&self) -> Self::Instant;
```

Comment:
I don't understand the difference between `on_request` and `on_call`.

is `on_request`  intended for `JSON-RPC notifications and method call`? When is this useful to have?

Compared to `on_call`, `on_notif`?

Reply:
This is mirroring `jsonrpc` which I believe treats batch arrays as single requests. It does logically make sense to differentiate the two, and it makes batches easier to manage because you don't have to get a timestamp for all calls in the batch (which would be the same). All that said, I'm happy with changes to labeling here.

Reply:
As I understood it:
`on_request`: all incoming requests, regardless of what kind (call, notification, subscription, unsubscription…); for instance a batch request would count as one request
`on_call`: actual method calls; each call in a batch would count as one call

Reply:
that makes sense, the naming is alright but the documentation could be a little more detailed.

batch requests can contain notifications but that isn't a problem here right?

Reply:
> batch requests can contain notifications but that isn't a problem here right?

We don't support client->server notifications yet so that's not really a thing :upside_down_face:. The notifications sent from the server as a part of subscription aren't tracked by middleware, there isn't really a way to make it so without sticking middleware into modules, and that creates a huge mess.

Reply:
yupp, right we don't even try to decode `batches` with notifications good point.